s4e-backend/src/main/java/pl/cyfronet/s4e/service/InvitationService.java

```diff
@@ -0,0 +1,115 @@
+    private final TemplateEngine templateEngine;
+    private final MailService mailService;
+    private final MailHelper mailHelper;
+
+    private static final TemporalAmount EXPIRE_IN = Duration.ofDays(7);
+
+    @Transactional(rollbackFor = InvitationCreationException.class)
+    public Invitation createInvitationFrom(CreateInvitationRequest request) throws Exception {
+        var institution = institutionRepository.findBySlug(request.getInstitutionSlug(), Institution.class)
+            .orElseThrow(() -> new InvitationCreationException("Institution doesn't exists"));
+        var invitation = invitationRepository.findByEmailAndInstitutionId(request.getEmail(), institution.getId());
+        if (invitation != null) {
+            throw new InvitationCreationException("Invitation already exists in DB");
+        }
+
+        var verificationToken = UUID.randomUUID().toString();
+        invitation = Invitation.builder()
+            .email(request.getEmail())
+            .token(verificationToken)
+                .status(InvitationStatus.WAITING)
+            .institution(institution)
+            .build();
+        invitationRepository.save(invitation);
+
+        return invitation;
+    }
+
+    public void sendInvitation(Invitation invitation, Locale locale) {
+        Object[] subjectInstitution = new Object[]{invitation.getInstitution().getName()};
+        String subject = messageSource
+            .getMessage(
+                "email.invitation.subject {0}",
+                subjectInstitution,
+                locale
+            );
+        String confirmUrl = mailHelper.prefixWithDomain("/login?token=" + invitation.getToken());
+        String rejectUrl = mailHelper.prefixWithDomain("/login?reject=''&token=" + invitation.getToken());
+
+        Context ctx = new Context(locale);
+        mailHelper.injectCommonVariables(ctx);
+
+        ctx.setVariable("email", invitation.getEmail());
+        ctx.setVariable("institutionName", invitation.getInstitution().getName());
+        ctx.setVariable("confirmUrl", confirmUrl);
+        ctx.setVariable("rejectUrl", rejectUrl);
+
+        String plainText = templateEngine.process("invitation-email.txt", ctx);
+        String htmlText = templateEngine.process("invitation-email.html", ctx);
+
+        mailService.sendEmail(invitation.getEmail(), subject, plainText, htmlText);
+    }
+
+    public Institution confirmInvitation(ConfirmInvitationRequest request, AppUser user) throws InvitationConfirmException {
+        var invitation = invitationRepository.findByToken(request.getToken(), Invitation.class)
+                .orElseThrow(() -> new InvitationConfirmException("Invitation with token doesn't exists"));
+
+        var institution = invitation.getInstitution();
+        addMemberToDefaultGroup(user, institution);
+
+        return institution;
+    }
+
+    private void addMemberToDefaultGroup(AppUser user, Institution institution) throws InvitationConfirmException {
```

Comment:
i would move this function to GroupService or there is method GroupService::addMember, why not use it?

Reply:
```
Daniel Kryska  9:23 AM
+ myślałem o tej funkcji, żeby dodać użytkownika do defaultowej grupy
9:23
addToGroup

Jakub Sawicki  9:25 AM
w jakiej klasie to addToGroup jest?

Daniel Kryska  9:25 AM
GroupService
9:25
A z tymi detalami to używacie tego przez serwis
9:25
AppUserDetails appUserDetails = AppUserDetailsSupplier.get();
        var user = appUserService.findByEmailWithRolesAndGroupsAndInstitution(appUserDetails.getUsername(), AppUserResponse.class)
                .orElseThrow(() -> new NotFoundException("User not found for email: '" + appUserDetails.getUsername() + "'"));

Jakub Sawicki  9:28 AM
ale GroupService::addToGroup jest prywatne i tak

Daniel Kryska  9:30 AM
co za problem zrobić go publicznym?

Jakub Sawicki  9:31 AM
z jakiegoś powodu jest prywatny

Daniel Kryska  9:31 AM
Ogólnie szukam rozwiązania problemu jakim jest dodanie użytkownika do instytucji

Jakub Sawicki  9:31 AM
można by dodać nową metodę, do tego

Daniel Kryska  9:31 AM
Bo poprzednia implementacja była mockiem
9:31
Masz na myśli do tego serwisu czy do serwisu invitation?
9:32
do tego = groupService

Jakub Sawicki  9:32 AM
myślałem o tym, że do tego = GroupService
9:34
GroupService::addToGroup ma sygnaturę: (Set<UserRole> rolesFromRequest, AppUser user, Set<UserRole> userRoles)
9:35
ja bym w sumie olał może już ten GroupService jednak, i zrobił na razie implementację z zaszytą grupą __default__  w InvitationService
9:35
a jak będzie refactor pod kątem wywalenia grup to się to ewentualnie ujednolici
9:35
i nie dodawałbym zależności do GroupService
```

Reply:
I'd keep it as I advised Daniel. This functionality will be refactored soon anyway, and in this way you avoid interdependency between services.